payara/AutoScale-Groups
# FISH-1224 SSH Nodes Scaling Group Extension

Creates an extension for scaling up and down on SSH nodes. This also includes a switch over to Community, changing the payara.version property and adjusting the snapshot/deployment repositories.

To use, build and then copy the following artefacts into the _payara5/glassfish/modules_ directory:
* autoscale-groups-api.jar
* autoscale-groups-core.jar
* autoscale-groups-console.jar
* autoscale-groups-nodes-plugin-core.jar
* autoscale-groups-nodes-plugin-console.jar

### Nodes Scaler

* New commands:
  * `create-nodes-scaling-group` - Creates a Nodes Scaling Group with the supplied config. 
    * Cannot be configured to use `default-config` or `server-config`
    * Must be attached to a deployment group
    * Must have at least one SSH node set (also accepts the default "DAS" node localhost-xyz)
  * `set-nodes-scaling-group-configuration` - Changes the configuration of a nodes scaling group
  * `get-nodes-scaling-group-configuration` - Gets the configuration of a nodes scaling group

The nodes scaler attempts to balance the number of instances across its available nodes when scaling up and down. Instances chosen for scaling down is random outside of the nodes balancing criteria, it does not currently look at whether instances are under load or even running.

Admin Console integration is a new tab available under _Deployment Groups > AutoScale Groups > Nodes Scaling Groups_. From here you can see the configured nodes scaling groups, create new ones, edit existing ones, and delete one or more.

### Additional Changes
* Adjusts the API in a number of ways (e.g. some interfaces became abstract classes) so that there's some shared validation and commands.
  * A `Scales` annotation was added to help link up a scaler _service_ to a scaling group _type_ (e.g. `NodesScaler` is the scaler service for a `NodesScalingGroup` configbean).
* The validation was moved from the ConfigBean to the admin commands, since I was having no end of trouble getting it to work properly.
* "Core" Admin Console integration added, displaying the active scaling group for a deployment group, and supplying a means of tabs for various impl extentions to be added.
* Added a "Core" module in addition to the API, containing generic admin commands and impl/util classes.
  * New commands:
    * `delete-scaling-group` - Deletes a scaling group, regardless of its type
    * `get-deployment-group-scaling-group` - Returns the scaling group attached to a deployment group
    * `list-scaling-groups` - Returns all configured scaling groups (regardless of type)
    * `scale-up` - Scales up a deployment group using its configured scaling group (create and start instances)
    * `scale-down` - Scales down a deployment group using its configured scaling group (stop and delete instances)
  * A ScaleCommandHelper util class which contains altered versions of various methods found in the `ClusterCommandHelper` class

### Testing

Basic Workflow:
* Create config - `asadmin copy-config default-config gruppy-config`
* Create deployment group - `asadmin create-deployment-group gruppy`
* Create remote SSH node - `asadmin create-node-ssh --nodehost xyz --installdir xyz --sshuser xyz --install noddy1`
* Create a second remote SSH node - `asadmin create-node-ssh --nodehost ijk --installdir xyz --sshuser xyz --install noddy2`
* Create nodes scaling group - `asadmin create-nodes-scaling-group --nodes "localhost-domain1,noddy1,noddy2! --config gruppy-config --deploymentgroup gruppy scaley`
* Scale up deployment group by 3 - `asadmin scale-up -q 3 gruppy`
  * Each node should have 1 instance created on it and started (assuming an equal balance of nodes beforehand)
* Scale down deployment group by 3 - `asadmin scale-down -q 3 gruppy`
  * All instances should be stopped and removed.

Of note here - depending on the speed of response, it's possible for a command to return from the `stop-instance` command before the instance has _fully_ stopped, which means when the follow-up delete command is executed it doesn't delete the instance properly (removed from config but not from filesystem).

Configuration Validation:
* Check that you can't configure two scaling groups to use the same deployment group
* Check that you can't create two scaling groups with the same name
* Check that you can't configure a scaling group to use a non-existent deployment group
* Check that you can't configure a scaling group to use a non-existent node
* Check that you can't configure a scaling group to use a non-existent config

Admin Console:
* Create config
* Create deployment group
* Go to AutoScale tab
* Go to Nodes AutoScale tab
* Create new
  * Use created config
  * Use created deployment group
  * Use localhost-domain1 as node
* Go back to AutoScale tab
* Scale up by 2
* Scale down by 2
